agent/app/agent.go

```diff
@@ -0,0 +1,404 @@
+	acceptInsecureCert *bool) (agent, error) {
+
+	ec2MetadataClient := ec2.NewEC2MetadataClient(nil)
+	if blackholeEC2Metadata {
+		ec2MetadataClient = ec2.NewBlackholeEC2MetadataClient()
+	}
+
+	log.Info("Loading configuration")
+	cfg, err := config.NewConfig(ec2MetadataClient)
+	if err != nil {
+		// All required config values can be inferred from EC2 Metadata,
+		// so this error could be transient.
+		log.Criticalf("Error loading config: %v", err)
+		return nil, err
+	}
+	cfg.AcceptInsecureCert = aws.BoolValue(acceptInsecureCert)
+	if cfg.AcceptInsecureCert {
+		log.Warn("SSL certificate verification disabled. This is not recommended.")
+	}
+	log.Debugf("Loaded config: %s", cfg.String())
+
+	dockerClient, err := engine.NewDockerGoClient(dockerclient.NewFactory(cfg.DockerEndpoint), cfg)
+	if err != nil {
+		// This is also non terminal in the current config
+		log.Criticalf("Error creating Docker client: %v", err)
+		return nil, err
+	}
+
+	return &ecsAgent{
+		ctx:               ctx,
+		ec2MetadataClient: ec2MetadataClient,
+		cfg:               cfg,
+		dockerClient:      dockerClient,
+		// We instantiate our own credentialProvider for use in acs/tcs. This tries
+		// to mimic roughly the way it's instantiated by the SDK for a default
+		// session.
+		credentialProvider: defaults.CredChain(defaults.Config(), defaults.Handlers()),
+	}, nil
+
+}
+
+func (agent *ecsAgent) printVersion() int {
+	version.PrintVersion(agent.dockerClient)
+	return exitcodes.ExitSuccess
+}
+
+func (agent *ecsAgent) start() int {
+	sighandlers.StartDebugHandler()
+
+	containerChangeEventStream := eventstream.NewEventStream(containerChangeEventStreamName, agent.ctx)
+	credentialsManager := credentials.NewManager()
+	state := dockerstate.NewTaskEngineState()
+	imageManager := engine.NewImageManager(agent.cfg, agent.dockerClient, state)
+
+	return agent.doStart(containerChangeEventStream, credentialsManager, state, imageManager)
+}
+
+func (agent *ecsAgent) doStart(containerChangeEventStream *eventstream.EventStream,
+	credentialsManager credentials.Manager,
+	state dockerstate.TaskEngineState,
+	imageManager engine.ImageManager) int {
+
+	taskEngine, currentEC2InstanceID, err := agent.newTaskEngine(containerChangeEventStream,
+		credentialsManager, state, imageManager)
+	if err != nil {
+		return exitcodes.ExitTerminal
+	}
+
+	stateManager, err := agent.newStateManager(taskEngine,
+		&agent.cfg.Cluster, &agent.containerInstanceArn, &currentEC2InstanceID)
+	if err != nil {
+		log.Criticalf("Error creating state manager: %v", err)
+		return exitcodes.ExitTerminal
+	}
+
+	client := ecsclient.NewECSClient(agent.credentialProvider, agent.cfg, agent.ec2MetadataClient)
+	err = agent.registerContainerInstance(taskEngine, stateManager, client)
+	if err != nil {
+		if isNonTerminal(err) {
+			return exitcodes.ExitError
+		}
+		return exitcodes.ExitTerminal
+	}
+
+	// Begin listening to the docker daemon and saving changes
+	taskEngine.SetSaver(stateManager)
```

Comment:
because naming is hard. The basic state manager is also a saver as it implements the `statemanager.Saver` interface.